metadata-io/src/main/java/com/linkedin/metadata/service/TagService.java

```diff
@@ -0,0 +1,455 @@
+
+    changes.addAll(entityProposals);
+    changes.addAll(schemaFieldProposals);
+
+    return changes;
+  }
+
+  @VisibleForTesting
+  List<MetadataChangeProposal> buildAddTagsToEntityProposals(
+      List<Urn> tagUrns,
+      List<ResourceReference> resources,
+      Authentication authentication
+  ) throws URISyntaxException {
+    final Map<Urn, GlobalTags> tagsAspects = getTagsAspects(
+        resources.stream().map(ResourceReference::getUrn).collect(Collectors.toSet()),
+        new GlobalTags(),
+        authentication
+    );
+
+    final List<MetadataChangeProposal> changes = new ArrayList<>();
+    for (ResourceReference resource : resources) {
+      GlobalTags globalTags = tagsAspects.get(resource.getUrn());
+      if (globalTags == null) {
+        continue; // Something went wrong.
+      }
+      if (!globalTags.hasTags()) {
+        globalTags.setTags(new TagAssociationArray());
+      }
+      addTagsIfNotExists(globalTags, tagUrns);
+      MetadataChangeProposal proposal = buildMetadataChangeProposal(
+          resource.getUrn(),
+          Constants.GLOBAL_TAGS_ASPECT_NAME,
+          globalTags
+      );
+      changes.add(proposal);
+    }
+    return changes;
+  }
+
+  @VisibleForTesting
+  List<MetadataChangeProposal> buildAddTagsToSubResourceProposals(
+      final List<Urn> tagUrns,
+      final List<ResourceReference> resources,
+      final Authentication authentication
+  ) throws URISyntaxException {
+
+    final Map<Urn, EditableSchemaMetadata> editableSchemaMetadataAspects = getEditableSchemaMetadataAspects(
+        resources.stream().map(ResourceReference::getUrn).collect(Collectors.toSet()),
+        new EditableSchemaMetadata(),
+        authentication
+    );
+
+    final List<MetadataChangeProposal> changes = new ArrayList<>();
+    for (ResourceReference resource : resources) {
+
+      EditableSchemaMetadata editableSchemaMetadata = editableSchemaMetadataAspects.get(resource.getUrn());
+      if (editableSchemaMetadata == null) {
+        continue; // Something went wrong.
+      }
+
+      EditableSchemaFieldInfo editableFieldInfo = getFieldInfoFromSchema(editableSchemaMetadata, resource.getSubResource());
+
+      if (!editableFieldInfo.hasGlossaryTerms()) {
+        editableFieldInfo.setGlossaryTerms(new GlossaryTerms());
+      }
+
+      addTagsIfNotExists(editableFieldInfo.getGlobalTags(), tagUrns);
+      changes.add(buildMetadataChangeProposal(resource.getUrn(), Constants.EDITABLE_SCHEMA_METADATA_ASPECT_NAME,
+          editableSchemaMetadata));
+    }
+
+    return changes;
+  }
+
+  @VisibleForTesting
+  List<MetadataChangeProposal> buildRemoveTagsToEntityProposals(
+      List<Urn> tagUrns,
+      List<ResourceReference> resources,
+      Authentication authentication
+  ) {
+    final Map<Urn, GlobalTags> tagsAspects = getTagsAspects(
+        resources.stream().map(ResourceReference::getUrn).collect(Collectors.toSet()),
+        new GlobalTags(),
+        authentication
+    );
```

Comment:
this case would be a perfect use case for patch mcps

Reply:
yessss! we will sub that in once its available